staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -735,16 +736,27 @@ func validateCustomResourceDefinitionValidation(customResourceValidation *apiext
 	return allErrs
 }
 
+type schemaTree struct {
+	MaxLength int64
+	Parent    *schemaTree
+	Schema    *apiextensions.JSONSchemaProps
+}
+
 var metaFields = sets.NewString("metadata", "kind", "apiVersion")
 
 // ValidateCustomResourceDefinitionOpenAPISchema statically validates
-func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSchemaProps, fldPath *field.Path, ssv specStandardValidator, isRoot bool, opts *validationOptions) field.ErrorList {
+func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSchemaProps, fldPath *field.Path, ssv specStandardValidator, isRoot bool, opts *validationOptions, parentNode *schemaTree) field.ErrorList {
 	allErrs := field.ErrorList{}
 
 	if schema == nil {
 		return allErrs
 	}
 
+	treeNode := &schemaTree{
+		Parent: parentNode,
+		Schema: schema,
+	}
+
```

Comment:
Would it be possible to instead, compute the "cardinality" at each level of depth and then passing that down? E.g.

```
ValidateCustomResourceDefinitionOpenAPISchema(..., cardinality)
cardinality = <this schema node's maxLength> * <cardinality of parent>
ValidateCustomResourceDefinitionOpenAPISchema(..., cardinality)
```